projects/output-and-error-handling/client/main_test.go

```diff
@@ -0,0 +1,84 @@
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"strconv"
+	"testing"
+	"time"
+)
+
+func TestHandleWeatherRequest_Success(t *testing.T) {
+	expRespSlice := [][]byte{[]byte("Today it will be sunny!"), []byte("Tomorrow it will be rainy!")}
+
+	for i, expected := range expRespSlice {
+		t.Run(fmt.Sprintf("Case %d", i+1), func(t *testing.T) {
+			svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusOK)
+				w.Write(expected)
+			}))
+			defer svr.Close()
+			client := &http.Client{}
+			resp, err := handleWeatherRequest(svr.URL, client, time.Now(), 3)
+			if err != nil {
+				t.Errorf("expected err to be nil got %v", err)
+			}
+
+			if resp != string(expected) {
+				t.Errorf("expected response body to be %s got %s", expected, resp)
+			}
+		})
+	}
+}
+
+func TestHandleWeatherRequest_InternalServerError(t *testing.T) {
+	errorMessage := "Internal Server Error"
+	errorStatusCode := http.StatusInternalServerError
+	svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(errorStatusCode)
+		w.Write([]byte(errorMessage))
+	}))
+	defer svr.Close()
+
+	client := &http.Client{}
+	_, err := handleWeatherRequest(svr.URL, client, time.Now(), 3)
+	if err == nil {
+		t.Error("expected error, got nil")
+	}
+
+	expectedError := fmt.Sprint(errorStatusCode, " : ", errorMessage)
+	if err.Error() != expectedError {
+		t.Errorf("expected error message to be %s, got %s", expectedError, err.Error())
+	}
+}
+
+func TestHandleRateLimited(t *testing.T) {
+	testCases := []int{3, 6, -1}
```

Comment:
I'd probably pull this out as a table-driven test so that the expectations are alongside the input values:

```go
type testCase struct {
    headerValue string
    wantValue time.Duration
    wantError error
}

testCases := []testCase{
    testCase{
        headerValue: "3",
        wantValue: 3 * time.Second,
        wantError: nil
    }
}
...
```

This avoids having so much logic in your test, and having to change the logic in the test if the input test-cases change.

Also, I'd probably mostly test this behaviour via a unit test of `convertTime`, and maybe just have one test that tests this via an actual HTTP request, rather than exclusively testing it via HTTP requests.

Reply:
> Also, I'd probably mostly test this behaviour via a unit test of `convertTime`, and maybe just have one test that tests this via an actual HTTP request, rather than exclusively testing it via HTTP requests.

If I have to test via `convertTime` function only how can I test the other error messages ? e.g the error on the `HandleRateLimited` ?

Reply:
Either with a separate test, or deciding not to bother :)

We don't need to just write one or two tests to try to cover all behaviours of our program, we can write lots of separate tests for different aspects of functionality.